### ISSUE TITLE
Could not authorize with Spotify [SOLUTION]

### ISSUE THREAD

**Issue by FDrico** [olejon/spotcommander#102]
**NOTE FROM DEVELOPER:** [Check my last comment for solution](https://github.com/olejon/spotcommander/issues/102#issuecomment-218767777).

First of all, I really appreciate your work. Simply amazing.

After reinstalling my system, I obviously had to authorize with Spotify. I installed the last fully supported version of Spotify, the necessary libraries, and Spotcommander on my phone. I checked on every item listed in the troubleshooting page. Spotify works fine.

When i enter http://localhost/spotcommander/, it asks me to authorize with Spotify. I clic on the button on the top right, which takes me to the authorization page in spotify.com, and I obviously clic OKAY.

This brings me back to http://localhost/spotcommander, with the sign: "Could not authorize with Spotify. Tap the top right icon to try again".

I've done this also from my phone, with the same result.

Any ideas?

Thank you

**Comment by olejon** [comment]
OK, I just tried and it works fine here. BTW: Which distro are you running, and do you have a Facebook or regular account?

So some things to check:

* When you are sent back to SpotCommander and you get "Could not authorize", can you copy/paste the URL in the address bar here?
* You can try going to www.spotify.com and log out from there, then try to authorize in SpotCommander
* You can try to authorize from an `incognito window` in Google Chrome for instance, to avoid extensions from doing something
* If running a Ubuntu/Debian-based system, check `/var/log/apache2/error.log` after you get the error, or if your system only logs to systemctl it would be something like `sudo journalctl --no-pager -b -u httpd` or `sudo journalctl --no-pager -b -u apache` or `sudo journalctl --no-pager -b -u apache2`

**Comment by FDrico** [comment]
Hi!

I'm using Xubuntu 15.10, and i'm using a regular account, without weird character in the name (fdrico).

The URL in the address bar where SpotCommander takes me to is:

`https://accounts.spotify.com/en/authorize?client_id=854ed72e47fe4e3c89d5276ac2c47ceb&response_type=code&redirect_uri=http:%2F%2Fwww.olejon.net%2Fcode%2Fspotcommander%2Fapi%2F1%2Fspotify%2Fcallback%2F&state=1462787634030&scope=user-read-private%20playlist-read-private%20playlist-read-collaborative%20playlist-modify-public%20playlist-modify-private%20user-follow-modify%20user-follow-read%20user-library-read%20user-library-modify%20user-top-read&show_dialog=true`

And the URL in the address bar when it takes me to SpotCommander is this:

`http://localhost/spotcommander/#profile//spotify_token=eyJhY2Nlc3NfdG9rZW4iOiJCUUFscUhXV3B4VHozNXZ3Z0puNzczc2ZPYjVKWVg2SjZyb2IyYjBRZDRQa3JQdkZJRVhXWjc3NzVxSm5hdWVZS3NsbHBDVDBCUUc3WUVnZUFuUFVIUmtPbGZqNnR6ck8tekZ4aHBDQm44UHVhRDczQjl4MlIza1RXbHJLQUJ5cEozZGVTR1VXSXhVQkhOWkRWSXV3R2ZLYlVDZTBuS3ZHTGp3X1VJdG9tZzJZbnM5Nlhxem84WUNBMFFNYlh3QmNnS0xHV3I5OWxCQzRkbURVQ0JnNEE0SVVRSUx6OVNxUlotWENlbW5HNEZnc0dDQU5TR1BuemEwcnozcUtrMFFlekpFQTAzXzkyZ0pNbnc2WnZzbkJMUy1fS2tLOFFlamFJOEpqcnciLCJ0b2tlbl90eXBlIjoiQmVhcmVyIiwiZXhwaXJlc19pbiI6MzYwMCwicmVmcmVzaF90b2tlbiI6IkFRQ2JEUE5WSDI2angyTkpvczNVdVNqUnNjMThicXFBNE4wWjNvRDV5VjhZNDEySWdJbGxZQ0dUbXhhZ3RTUkFWMWpDNExCWklPNTAzeTdrOEo2UGNMNGJhQzk5cVFKQWVSX0V4djNoWXNRM29ZUHFUWnprRWlPWWJ6X1oxNHNfbGYwIn0=/0`

- I've tried using Chromium (had to install it for this), so I wasn't logged in. When i am logged in, the same thing happens. I also tried Firefox.

- I just tried that. Same thing happens.

- `/var/log/apache2/error.log` says this:

```
[Mon May 09 06:39:14.207334 2016] [mpm_prefork:notice] [pid 1134] AH00163: Apache/2.4.12 (Ubuntu) configured -- resuming normal operations
[Mon May 09 06:39:14.207382 2016] [core:notice] [pid 1134] AH00094: Command line: '/usr/sbin/apache2'
```

- `sudo journalctl --no-pager -b -u apache2` is the only one with entries, and it says, after trying to authorize:

```
may 09 06:31:17 fede-HP apache2[998]: *
may 09 06:31:17 fede-HP systemd[1]: Started LSB: Apache2 web server.
may 09 06:39:14 fede-HP systemd[1]: Reloading LSB: Apache2 web server.
may 09 06:39:14 fede-HP apache2[1616]: * Reloading web server apache2
may 09 06:39:14 fede-HP apache2[1616]: *
may 09 06:39:14 fede-HP systemd[1]: Reloaded LSB: Apache2 web server.
```

Nothing changes there.

I also reinstalled SpotCommander, after removing and purging, but nothing happened.

**Comment by olejon** [comment]
Ok, no Apache errors. So all the stuff you see in the last URL after `spotify_token=` until `/` is Base64 encoded. When I decode it I get:

`{"access_token":"BQAlqHWWpxTz35vwgJn773sfOb5JYX6J6rob2b0Qd4PkrPvFIEXWZ7775qJnaueYKsllpCT0BQG7YEgeAnPUHRkOlfj6tzrO-zFxhpCBn8PuaD73B9x2R3kTWlrKABypJ3deSGUWIxUBHNZDVIuwGfKbUCe0nKvGLjw_UItomg2Yns96Xqzo8YCA0QMbXwBcgKLGWr99lBC4dmDUCBg4A4IUQILz9SqRZ-XCemnG4FgsGCANSGPnza0rz3qKk0QezJEA03_92gJMnw6ZvsnBLS-_KkK8QejaI8Jjrw","token_type":"Bearer","expires_in":3600,"refresh_token":"AQCbDPNVH26jx2NJos3UuSjRsc18bqqA4N0Z3oD5yV8Y412IgIllYCGTmxagtSRAV1jC4LBZIO503y7k8J6PcL4baC99qQJAeR_Exv3hYsQ3oYPqTZzkEiOYbz_Z14s_lf0"}`

You can paste it in here to see it more clearly, then click Viewer at the top left:

http://jsonviewer.stack.hu/

Everything looks normal. A token only lasts 3600 seconds as you can see, so I can not check it.

Can you check it here:

https://developer.spotify.com/web-api/console/get-current-user/

Click the green GET OAUTH TOKEN button, then mark every checkbox, even those not relevant, just to be sure. Then click the green TRY IT button, and you should get a request and a response. I am interested in the JSON response. If it looks okay.

**Comment by FDrico** [comment]
Ok, i did what you said. The response is this:

```
HTTP/1.1 200 OK
Access-Control-Allow-Headers: Accept, Authorization, Origin, Content-Type
Access-Control-Max-Age: 604800
X-Content-Type-Options: nosniff
Content-Encoding: gzip
Transfer-Encoding: chunked
Access-Control-Allow-Credentials: true
Strict-Transport-Security: max-age=31536000;
Vary: Authorization
Keep-Alive: timeout=600
Server: nginx
Connection: keep-alive
Cache-Control: private, max-age=0
Date: Tue, 10 May 2016 10:11:48 GMT
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE
Content-Type: application/json; charset=utf-8
```

`{
  "birthdate" : "1982-01-06",
  "country" : "AR",
  "display_name" : null,
  "email" : "my-email@gmail.com",
  "external_urls" : {
    "spotify" : "https://open.spotify.com/user/fdrico"
  },
  "followers" : {
    "href" : null,
    "total" : 1
  },
  "href" : "https://api.spotify.com/v1/users/fdrico",
  "id" : "fdrico",
  "images" : [ ],
  "product" : "free",
  "type" : "user",
  "uri" : "spotify:user:fdrico"
}`

**Comment by olejon** [comment]
Hm, this is really hard to understand. I can only find 1 log entry that ended in "Fail". Is your public IP address `xxx.193.128.21` ?

**Comment by olejon** [comment]
Read above first.

BTW: Is there anything in the file `/usr/share/spotcommander/run/spotify.token` ?

**Comment by FDrico** [comment]
Hi!

`xxx.193.128.21` is indeed my IP. I don't know why it appears only once. I tried many, many times.

When i get back home, i will be able to answer your second question.

**Comment by FDrico** [comment]
There is nothing in the file `usr/share/spotcommander/run/spotify.token`, i just checked.

**Comment by olejon** [comment]
This is really weird. Could you try to create a new free account and try to authorize with it just to check? Simple username and password.

**Comment by FDrico** [comment]
I just created another username. Tried and, sadly, it doesn't work either. Same problems. Of course i tried with chromium in private mode.

I also reinstalled spotify, just to be sure.

I agree, this is weird.

But, even if we don't get to the bottom of this, thank you very much for your time. You have lots of patience. ¡Gracias!

**Comment by olejon** [comment]
Alright, I found the error. It was the introduction of the new feature "Top Tracks" in your profile. If you have a relatively new account there will be no "Top Tracks", and I made a code mistake.

To fix this, replace the code on line `2067` in the file `/usr/share/spotcommander/functions.php` with this:

`if(!empty($profile['id']) && isset($profile_top_tracks['items']))`

Will release a new version soon and keep this open for others to see.
